test/integration/controllermanager/managedseedset/managedseedset_test.go

```diff
@@ -0,0 +1,376 @@
+			makeReplicaReady()
+
+			Eventually(func(g Gomega) {
+				g.Expect(testClient.Get(ctx, client.ObjectKeyFromObject(managedSeedSet), managedSeedSet)).To(Succeed())
+				g.Expect(managedSeedSet.Status.PendingReplica).To(BeNil())
+				g.Expect(managedSeedSet.Status.ReadyReplicas).To(Equal(int32(1)))
+			}).Should(Succeed())
+		})
+	})
+
+	Context("scale-out", func() {
+		It("should create new replica and update the status.replicas field because spec.replicas value was increased", func() {
+			Eventually(func(g Gomega) {
+				g.Expect(testClient.Get(ctx, client.ObjectKeyFromObject(managedSeedSet), managedSeedSet)).To(Succeed())
+				g.Expect(managedSeedSet.Status.Replicas).To(Equal(int32(1)))
+			}).Should(Succeed())
+
+			By("Update the Replicas to 2")
+			patch := client.MergeFrom(managedSeedSet.DeepCopy())
+			managedSeedSet.Spec.Replicas = pointer.Int32(2)
+			Expect(testClient.Patch(ctx, managedSeedSet, patch)).To(Succeed())
+
+			By("Make one replica ready, to enable controller to pick the next replica")
+			makeReplicaReady()
+
+			Eventually(func(g Gomega) {
+				g.Expect(testClient.Get(ctx, client.ObjectKeyFromObject(managedSeedSet), managedSeedSet)).To(Succeed())
+				g.Expect(managedSeedSet.Status.ReadyReplicas).To(Equal(int32(1)))
+				g.Expect(managedSeedSet.Status.Replicas).To(Equal(int32(2)))
+			}).Should(Succeed())
+		})
+	})
+
+	Context("scale-in", func() {
+		It("should delete replicas and update the status.replicas field because spec.replicas value was updated with lower value", func() {
+			Eventually(func(g Gomega) {
+				g.Expect(testClient.Get(ctx, client.ObjectKeyFromObject(managedSeedSet), managedSeedSet)).To(Succeed())
+				g.Expect(managedSeedSet.Status.Replicas).To(Equal(int32(1)))
+			}).Should(Succeed())
+
+			By("Scale-up the Replicas to 2")
+			patch := client.MergeFrom(managedSeedSet.DeepCopy())
+			managedSeedSet.Spec.Replicas = pointer.Int32(2)
+			Expect(testClient.Patch(ctx, managedSeedSet, patch)).To(Succeed())
+
+			By("Make one replica ready, to enable controller to pick the next replica")
+			makeReplicaReady()
+
+			Eventually(func(g Gomega) {
+				g.Expect(testClient.Get(ctx, client.ObjectKeyFromObject(managedSeedSet), managedSeedSet)).To(Succeed())
+				g.Expect(managedSeedSet.Status.ReadyReplicas).To(Equal(int32(1)))
+				g.Expect(managedSeedSet.Status.Replicas).To(Equal(int32(2)))
+			}).Should(Succeed())
+
+			By("Scale-down the Replicas to 1")
+			patch = client.MergeFrom(managedSeedSet.DeepCopy())
+			managedSeedSet.Spec.Replicas = pointer.Int32(1)
+			Expect(testClient.Patch(ctx, managedSeedSet, patch)).To(Succeed())
+
+			Eventually(func(g Gomega) {
+				g.Expect(testClient.Get(ctx, client.ObjectKeyFromObject(managedSeedSet), managedSeedSet)).To(Succeed())
+				g.Expect(managedSeedSet.Status.Replicas).To(Equal(int32(1)))
```

Comment:
Can you make assertions for `ReadyReplicas` and `PendingReplicas` here?

Reply:
`ReadyReplicas` are not checked yet

Reply:
Done here it seems ? https://github.com/gardener/gardener/blob/43c2a1915004bac8f1e0c6b13857d7f49208905e/test/integration/controllermanager/managedseedset/managedseedset_test.go#L325

Reply:
Yes, but shouldn't we assert this field again now that we have scaled down?

Reply:
I think what's missing here is an assertion that the ReadyReplicas equal 2 before the scale down. 

I am also curious how in the previous `Eventually` call (line 311) the `ReadyReplicas` are 1 and the `PendingReplica` is nil. Shouldn't the controller set `PendingReplica` to the second replica that has to be created as part of the scale up?

I think a check whether ReadyReplicas equals 2 is also missing in the scale-out test.

Reply:
Focus of the test for scale-up is to show that once the first replica is ready , the controller boots up the second replica and updates the `managedseedset.status.replicas` status to 2. Hence the assertions `ReadyReplica = 1` and second replica will as `pendingReplica`. After scale down the pending replica will deleted as the non ready replica will have lower priority and hence it will be picked up by controller for deletion during `scale-in`.So after scale-down only the ready replica will left and there will be no pending replica. Hope it clears the confusion now ?

Reply:
Yes thanks! I had incorrectly read the the `PendingReplica` assertion as `To(BeNil())` when it was actually `NotTo(BeNil())`.